apis/v1alpha1/httproute_types.go

```diff
@@ -427,64 +446,114 @@ type HTTPRequestHeaderFilter struct {
+type HTTPRouteForwardTo struct {
+	// ServiceName refers to the name of the Service to forward matched requests
+	// to. When specified, this takes the place of BackendRef. If both
+	// BackendRef and ServiceName are specified, ServiceName will be given
+	// precedence. If the referent cannot be found, controllers must set the
+	// "InvalidRoutes" status condition on any Gateway that includes this
+	// Route to true.
 	//
-	// Support: core
+	// Support: Core
 	//
-	// +kubebuilder:validation:MinItems=1
-	To []HTTPForwardToTarget `json:"to"`
+	// +optional
+	// +kubebuilder:validation:MaxLength=253
+	ServiceName *string `json:"serviceName,omitempty"`
 
-	// ExtensionRef is an optional, implementation-specific extension to the
-	// "action" behavior.  The resource may be "configmaps" (use the empty
-	// string for the group) or an implementation-defined resource (for
-	// example, resource "myrouteactions" in group "networking.acme.io").
-	// Omitting or specifying the empty string for both the resource and
-	// group indicates that the resource is "configmaps".  If the referent
-	// cannot be found, the "InvalidRoutes" status condition on any Gateway
-	// that includes the HTTPRoute will be true.
+	// BackendRef is a reference to a backend to forward matched requests to. If
+	// both BackendRef and ServiceName are specified, ServiceName will be given
+	// precedence. If the referent cannot be found, controllers must set the
+	// "InvalidRoutes" status condition on any Gateway that includes this Route
+	// to true.
 	//
-	// Support: custom
+	// Support: Extended
+	//
+	// +optional
+	BackendRef *LocalObjectReference `json:"backendRef,omitempty"`
+
+	// Port specifies the destination port number to use for the backend
+	// referenced by the ServiceName or BackendRef field. If unspecified and a
+	// Service object consisting of a single port definition is the backend,
+	// that port will be used. If unspecified and the backend is a Service
+	// object consisting of multiple port definitions, controllers must set the
+	// "InvalidRoutes" status condition on any Gateway that includes this Route
+	// to true.
+	//
+	// Support: Core
 	//
 	// +optional
-	ExtensionRef *RouteActionExtensionObjectReference `json:"extensionRef"`
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=65535
+	Port *int32 `json:"port,omitempty"`
+
+	// Weight specifies the proportion of traffic forwarded to the backend
+	// referenced by the ServiceName or BackendRef field. computed as
+	// weight/(sum of all weights in this ForwardTo list). Weight is not a
+	// percentage and the sum of weights does not need to equal 100. If only one
+	// backend is specified, 100% of the traffic is forwarded to that backend.
+	// If unspecified, weight defaults to 1.
+	//
+	// Support: Core (HTTPRoute)
+	//
+	// +kubebuilder:default=1
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=10000
+	Weight int32 `json:"weight,omitempty"`
```

Comment:
I'm not sure but `omitempty` shouldn't be present here?
Maybe this belongs to the clean up PR.

Reply:
Unfortunately every field with a default applied needs to be `omitempty` or `optional`. In this case `omitempty` seemed to make more sense here. Just tried this out to verify and got `missing required field "weight" in io.x-k8s.networking.v1alpha1.HTTPRoute.spec.rules.forwardTo`.

Reply:
One day, I'll get the hang of this. I need to sit down and get to the bottom of all the different possibilities here.
Apologies for the noise.

Reply:
No worries, I've been super confused as well on the appropriate combination of omitempty, optional, etc.